tests/test_subpixmaxima.py

```diff
@@ -157,22 +127,47 @@ def test_generate_keypoints_batched():
         imgaug_transform=imgaug_transform,
     )
     dl = DataLoader(dataset=dataset, batch_size=10)
-    SubPixMax = SubPixelMaxima(
-        output_shape=(96, 96),
-        output_sigma=torch.tensor(1.25, device=_TORCH_DEVICE),
-        upsample_factor=torch.tensor(2, device=_TORCH_DEVICE),
-        coordinate_scale=torch.tensor(4, device=_TORCH_DEVICE),  # 2 ** 2
-        confidence_scale=torch.tensor(1, device=_TORCH_DEVICE),
-        threshold=None,
-        device=_TORCH_DEVICE,
-    )
-    for batch in dl:
-        gt_keypts = batch["keypoints"].reshape(-1, 17, 2)
-        heatmaps = generate_heatmaps(
-            gt_keypts, height=384, width=384, output_shape=(96, 96)
+    for batch_dict in dl:
+        keypts_gt = batch_dict["keypoints"].reshape(-1, 17, 2)
+        heatmaps_torch = generate_heatmaps(
+            keypts_gt,
+            height=384,
+            width=384,
+            output_shape=(96, 96)
+        )
+
+        # find soft argmax and confidence of ground truth heatmap
+        softmaxes_gt = spatial_softmax2d(
+            batch_dict["heatmaps"].to(_TORCH_DEVICE),
+            temperature=torch.tensor(100).to(_TORCH_DEVICE))
+        preds_gt = spatial_expectation2d(
+            softmaxes_gt,
+            normalized_coordinates=False
+        )
+        confidences_gt = torch.amax(softmaxes_gt, dim=(2, 3))
+
+        # find soft argmax and confidence of generated heatmap
+        softmaxes_torch = spatial_softmax2d(
+            heatmaps_torch.to(_TORCH_DEVICE),
+            temperature=torch.tensor(100).to(_TORCH_DEVICE))
+        preds_torch = spatial_expectation2d(
+            softmaxes_torch,
+            normalized_coordinates=False
         )
-        gt_maxima, gt_confidence = SubPixMax.run(batch["heatmaps"].to(_TORCH_DEVICE))
-        torch_maxima, confidence_t = SubPixMax.run(heatmaps.to(_TORCH_DEVICE))
-        assert (gt_maxima == torch_maxima).all()
-        loss = MaskedMSEHeatmapLoss(batch["heatmaps"], heatmaps)
+        confidences_torch = torch.amax(softmaxes_torch, dim=(2, 3))
+
+        # og_maxima, gt_confidence = SubPixMax.run(heatmap_gt.to(_TORCH_DEVICE))
+        # torch_maxima, confidence_t = SubPixMax.run(heatmap_torch.to(_TORCH_DEVICE))
+        print(preds_gt[-1])
+        print(preds_torch[-1])
+        assert torch.allclose(preds_gt, preds_torch)
```

Comment:
that's a cool one! happy to learn it from you

Reply:
:)